src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventCounter.cs

```diff
@@ -42,7 +42,9 @@ public EventCounter(string name, EventSource eventSource) : base(name, eventSour
             {
                 bufferedValues[i] = UnusedBufferSlotValue;
             }
+#pragma warning disable CS8601 // Possible null reference assignment.
             Volatile.Write(ref _bufferedValues, bufferedValues);
+#pragma warning restore CS8601
```

Comment:
nit, in some places we are using the [`null forgiving operator`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) in situations like these. e.g.
```suggestion
            // ignore CS8601: Possible null reference assignment.
            Volatile.Write(ref _bufferedValues, bufferedValues!);
```

Reply:
Is this a compiler bug? I don't see why the warning is being issued here.

Reply:
Maybe it is because we have `[NotNullIfNotNull("value")] ref T location, T value` in `Volatile.Write<T>`. minimal repro could be:
```c#
using System;
using System.Runtime.CompilerServices;
using System.Diagnostics.CodeAnalysis;

class C
{
#nullable enable
    static void X<T>([NotNullIfNotNull("value")] ref T location, T value) where T : class
    { }

    private readonly double[] foo;

    C()
    {
        if (Environment.CommandLine.Length > 0) foo = new double[5];

        double[] bar = new double[3];
        X(ref foo, bar); // gives CS8601, with `ref foo!`, the warning disappears. 
    }
}
```

Reply:
@jcouv, @cston, is this a compiler bug?

Reply:
Thanks @am11 for the minimal repro. 

The `ref foo` argument contributes `double[]` to method type inference. But the incoming value of `foo` is maybe-null. Thus the warning.
We could argue that `ref foo` should contribute type `double[]?` instead, because the incoming value is maybe-null. That's not generally beneficial (one would get a warning on `ref` output), but it helps here because of the `NotNullIfNotNull` attribute. Please file a `roslyn` issue for follow-up.

Regardless of this type inference question, the declaration of `X`/`Volatile.Write<T>` must be adjusted.
With the current declaration, even a call with explicit type arguments would warn: 
`X<double[]?>(ref foo, bar); // warning CS8634: The type 'double[]?' cannot be used as type parameter 'T' in the generic type or method 'C.X<T>(ref T, T)'. Nullability of type argument 'double[]?' doesn't match 'class' constraint.`  ([example](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNGSAmIDUAfAAQCYBGAWACgCAGAAgJIDoAlAVwDsYBLMORgYQD2YAA5cANogDKiAG5cAxnCgBuSjXpMAIlwgBzdoNiKoAwZjgBBdhHEBPKF1WV1RWv0oBvSgGJ2rcXEIYElaOBsQuEpaGM0ANnoAFloADQAeABUAPgAKAG0AOUEYAoDxAEkAMyKSspyAIllbVjh6gEoAXVoEOEraDNpxQQUIbkF2JH7aJvEWttoAdwALRDgpkHoiaNjPWgBfFwpY2hEELib4brgITHH7WlvWSLyuysFBNQptmP4ctu/aN4jsdYlw+jkAKLseQIca8ThmMCQdiYAAyXHYfFR4T0MCWtCytGo8zegloAF5aJiFg9BE9JHkAKwdT4A46PZ5dYAQBAUqlwGkchkAZhZbNi6SFcBeAH5cj0+qTJtyEG0VLQAPQa2h6c7KdxSAAccWoJEmCy4eNoAAMFbRSQBCa2TPFrBY89gYvQPJwQEQia4IUy0AEHCh7IA===))

With the current signature, `value` is not-nullable so it is strange to have a `[NotNullIfNotNull("value")]` attribute... I would propose removing the `where T : class` constraint. Then an explicit call `X<double[]?>(ref foo, bar);` makes sense.

Depending if you want to wait for a roslyn fix or not, you could add `[AllowNull]` on the `location` parameter as a mitigation.

Reply:
@jcouv

>  I would propose removing the where T : class constraint

That would be functionally invalid: this method only works with reference types. Supplying an arbitrary value type would not work.

> With the current signature, value is not-nullable

Sure it is, if T is nullable. The constraint is `class?`.

Reply:
> We could argue that ref foo should contribute type double[]? instead, because the incoming value is maybe-null.

I would think so. Seems like ctors need to be special-cases here in the compiler, just as ctors have lots of other nullable-related special-cases (since they're then ones responsible for initializing non-nullable fields).

Reply:
> The constraint is `class?`.

Thanks for the clarification. I was working of the minimal repro, which has `where T : class` (possible typo).
Yes, `where T : class?` is good :-)

Reply:
I have opened https://github.com/dotnet/roslyn/issues/47663 . Also, I have deleted `Volatile.Write` here since it is not necessary. Publishing the counter has to issue the memory barrier in some form and other concrete DiagnosticCounter implementations depend on it.